common/lib/host_list_provider/monitoring/cluster_topology_monitor.ts

```diff
@@ -0,0 +1,482 @@
+    this.stopMonitoring = true;
+    this.hostMonitorsStop = true;
+    await Promise.all(this.runPromise ? this.untrackedPromises.concat(this.runPromise) : this.untrackedPromises);
+    this.untrackedPromises = [];
+    this.hostMonitors.clear();
+  }
+
+  async forceMonitoringRefresh(shouldVerifyWriter: boolean, timeoutMs: number): Promise<HostInfo[]> {
+    const currentHosts = this.topologyMap.get(this.clusterId);
+    if (currentHosts) {
+      return currentHosts;
+    }
+    if (shouldVerifyWriter) {
+      const client = this.monitoringClient;
+      this.monitoringClient = null;
+      this.isVerifiedWriterConnection = false;
+      await this.closeConnection(client);
+    }
+
+    return await this.waitTillTopologyGetsUpdated(timeoutMs);
+  }
+
+  async forceRefresh(client: any, timeoutMs: number): Promise<HostInfo[]> {
+    if (this.isVerifiedWriterConnection) {
+      return await this.waitTillTopologyGetsUpdated(timeoutMs);
+    }
+
+    // Otherwise use provided unverified connection to update topology
+    return await this.fetchTopologyAndUpdateCache(client);
+  }
+
+  async waitTillTopologyGetsUpdated(timeoutMs: number): Promise<HostInfo[]> {
+    const currentHosts: HostInfo[] = this.topologyMap.get(this.clusterId);
+    let latestHosts: HostInfo[];
+
+    // Notify monitor (that might be in delay) that topology should be updated.
+    this.requestToUpdateTopology = true;
+
+    if (timeoutMs === 0) {
+      return currentHosts;
+    }
+
+    let timeoutId: any;
+    const timeoutTask: Promise<void> = new Promise((resolve, reject) => {
+      timeoutId = setTimeout(() => {
+        reject("Topology update timed out.");
+      }, timeoutMs);
+    });
+
+    return await Promise.race([timeoutTask, await this.topologyUpdated])
+      .then((result) => {
+        latestHosts = this.topologyMap.get(this.clusterId);
+        return latestHosts === currentHosts ? null : latestHosts;
+      })
+      .catch((error: any) => {
+        logger.debug(`Could not update topology: ${error}.`);
+        if (JSON.stringify(error).includes("Topology update timed out.")) {
+          throw new AwsWrapperError(`ClusterTopologyMonitor topology update timed out in ${timeoutMs} ms.`);
+        }
+        return null;
+      })
+      .finally(async () => {
+        clearTimeout(timeoutId);
+      });
+  }
+
+  async fetchTopologyAndUpdateCache(client: any): Promise<HostInfo[]> {
+    if (!client) {
+      return null;
+    }
+
+    try {
+      const hosts: HostInfo[] = await this._hostListProvider.sqlQueryForTopology(client);
+      if (hosts) {
+        this.updateTopologyCache(hosts);
+      }
+      return hosts;
+    } catch (error) {
+      logger.error(`Error fetching topology: ${error}.`);
+    }
+    return null;
+  }
+
+  private async openAnyClientAndUpdateTopology() {
+    if (!this.monitoringClient) {
+      let client;
+      try {
+        client = await this._pluginService.forceConnect(this.initialHostInfo, this._properties);
```

Comment:
Need to connect using monitoring properties instead of the default properties. Similar comment for other calls to connect